src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,497 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+                                   SIZE_MAX),
+                  md_index, tl_mds[md_index].rsc.md_name);
+    }
+
+    result = UCS_PTR_BYTE_DIFF(buffer, p);
+
+out:
+    ucs_log_indent(-1);
+    return result;
+}
+
+static ucp_md_map_t ucp_find_md_map(ucp_context_h context,
+                                    const void *global_id_buf,
+                                    size_t global_id_size)
+{
+    ucp_md_map_t md_map = 0;
+    ucp_md_index_t md_index;
+    uct_md_attr_v2_t *md_attr;
+
+    for (md_index = 0; md_index < context->num_mds; md_index++) {
+        md_attr = &context->tl_mds[md_index].attr;
+
+        if ((ucp_memh_global_id_packed_size(md_attr) == global_id_size) &&
+            (memcmp(md_attr->global_id, global_id_buf, global_id_size) == 0)) {
+            md_map |= UCS_BIT(md_index);
+        }
+    }
+
+    return md_map;
+}
+
+static void
+ucp_memh_exported_tl_mkey_data_unpack(ucp_context_h context, 
+                                      const void **start_p,
+                                      const void **tl_mkey_buf_p,
+                                      ucp_md_map_t *md_map_p)
+{
+    const void *p = *start_p;
+    const void *next_tl_md_p, *end;
+    size_t tl_mkey_data_size;
+    size_t tl_mkey_size, global_id_size;
+    const void *tl_mkey_buf, *global_id_buf;
+    ucp_md_map_t md_map;
+
+    ucs_assert(p != NULL);
+
+    tl_mkey_data_size = ucp_memh_info_size_unpack(&p);
+    end               = UCS_PTR_BYTE_OFFSET(*start_p, tl_mkey_data_size);
+    ucs_assert(tl_mkey_data_size != 0);
+
+    tl_mkey_size = *ucs_serialize_next(&p, uint8_t);
+    ucs_assert(tl_mkey_size != 0);
+
+    tl_mkey_buf = ucs_serialize_next_raw(&p, void, tl_mkey_size);
+
+    global_id_size = *ucs_serialize_next(&p, uint8_t);
+    ucs_assert(global_id_size != 0);
+
+    global_id_buf = ucs_serialize_next_raw(&p, void, global_id_size);
+
+    /* Get local MD indices which corresponds to the remote ones */
+    md_map = ucp_find_md_map(context, global_id_buf, global_id_size);
+
+    next_tl_md_p = end;
+    ucs_assertv(p <= next_tl_md_p, "p=%p, next_tl_md_p=%p", p, next_tl_md_p);
+
+    *start_p       = next_tl_md_p;
+    *tl_mkey_buf_p = tl_mkey_buf;
+    *md_map_p      = md_map;
+}
+
+ucs_status_t
+ucp_memh_exported_unpack(ucp_context_h context, const void *export_mkey_buffer,
+                         ucp_memh_exported_unpacked_t *unpacked)
+{
+    const void *p = export_mkey_buffer;
+    const void *end;
+    uint16_t memh_info_size;
+    uint16_t UCS_V_UNUSED mem_info_parsed_size;
+    ucp_md_map_t local_md_map;
+    ucp_md_index_t remote_md_index;
+    ucp_md_index_t md_index;
+    const void *tl_mkey_buf;
+    ucp_memh_exported_tl_mkey_unpacked_t *tl_mkey;
+
+    ucs_assert(export_mkey_buffer != NULL);
+
+    /* Common memory handle information */
+    memh_info_size = ucp_memh_info_size_unpack(&p);
+    end            = UCS_PTR_BYTE_OFFSET(export_mkey_buffer, memh_info_size);
```

Comment:
"end" seems redundant, we can use memh_info_size

Reply:
done